StashiKu/aws-shop-be
# AWS-BE-Task-8: Integrate with SQL

### link to FE deploy - https://d3ar6ba8glakkn.cloudfront.net/
### link to PR for FE part - https://github.com/StashiKu/shop-angular-cloudfront/pull/7

####  **1.** Services that are `Fully integrated` with SQL and Front-End part:
 - `product-service`
 -  `import-service`
  **How to Test**
 - please see description for prev PR **[link](https://github.com/StashiKu/aws-shop-be/pull/5)**

#### **2.** Services that are `Partly integrated` with SQL and Front-End part:
 - `user service`
   **How to Test**
  - go to FE deploy
  - you will see a promt, it is an emulation of login process.
  - put your user name and it will be saved to the db and local storage. If you reload page, application will take user id from 
  local storage and check if it exists in db
 - to check if record really exists with your user name you can take **id** from the **local storage** and hit an endpoint
 `https://675cwvlc27.execute-api.eu-west-1.amazonaws.com/dev/api/users` with **query params** `userId: uuid`

My example
 `https://675cwvlc27.execute-api.eu-west-1.amazonaws.com/dev/api/users?userId=7c6c8e95-853e-434f-b254-b6fe49d5b17f`

#### **3.** Services that are `Partly integrated` with SQL and `are not integrated` with FE
- `order-service`
- `cart-service`
  **How to test**
[**A**] Cart service
 1. Scenario - user adds one item to the cart
  From data flow perspective we send `user data`, `product id` and `count` to BE, as a result the record in the `Carts` table is created with related record in the `Cart_Items`
a) hit this endpoint to get product list and pick up an `id` of any product
`https://ga15ffv8s8.execute-api.eu-west-1.amazonaws.com/dev/products`
b) hit endpoint **POST** https://675cwvlc27.execute-api.eu-west-1.amazonaws.com/dev/api/cart with `body`

```
{
    "user": {"id": your user id},
    "cartItem": {"productId": productId, "count": 2}
}
```
c) hit endpoint **GET** https://675cwvlc27.execute-api.eu-west-1.amazonaws.com/dev/api/car with `query params` like
`userId: id` to get cart

[**B**] Order Service
1. Order Creation
  a) hit endpoint **POST** https://675cwvlc27.execute-api.eu-west-1.amazonaws.com/dev/api/order  with `body`

```
{
    "userId": your user id,
    "status": "inProgress",
    "cartId":  id from prev test case,
    "payment": {
        "method": "credit card"
    },
    "delivery": {
        "address": "Mock",
        "first_name": "John",
        "last_name": "Doe"
    },
    "total": 13
}
```

  ### More info:
 -  PostgreSQL 
 - `Jest` is used for testing. Run `npm run test` to run tests
 - `TypeScript`
 - `Nest.js` 
 - `eslint`
 
**link to requirements** https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/8_integration_with_sql_database/task.md

### Evaluation criteria
[**1**] - Task 8.1 is implemented
[**3**] - Task 8.2 is implemented lambda links are provided and returns data
[**4**] - Task 8.3 is implemented lambda links are provided and products is stored in DB

**Additional (optional) tasks**
[**+1**] -  Create orders table and integrated with it Order model:
[**+1**] -  Create users table and integrate with it

### Total score - 6

**Thanks for reviewing** 👍 

### ⚠️ PS: if approve, please write a comment below "Approved"
